scipio/src/sys/mod.rs

```diff
@@ -185,54 +185,38 @@ impl From<Duration> for TimeSpec64 {
     }
 }
 
-/// Tasks interested in events on a source.
-#[derive(Debug)]
-pub(crate) struct Wakers {
-    /// Raw result of the operation.
-    pub(crate) result: Option<io::Result<usize>>,
-
-    /// Tasks waiting for the next event.
-    pub(crate) waiters: Vec<Waker>,
-}
-
-impl Wakers {
-    pub(crate) fn new() -> Self {
-        Wakers {
-            result: None,
-            waiters: Vec::new(),
-        }
-    }
-}
-
 /// A registered source of I/O events.
+#[derive(Debug)]
 pub struct InnerSource {
     /// Raw file descriptor on Unix platforms.
     raw: RawFd,
 
-    /// Tasks interested in events on this source.
-    wakers: RefCell<Wakers>,
-
     source_type: RefCell<SourceType>,
 
     io_requirements: IoRequirements,
 
-    enqueued: Cell<Option<EnqueuedSource>>,
+    state: RefCell<SourceState>,
 }
 
-pub struct EnqueuedSource {
-    pub(crate) id: SourceId,
-    pub(crate) queue: ReactorQueue,
+#[derive(Debug)]
+enum SourceState {
+    /// Neutral state, the Source either wasn't submitted yet, or was fully processed.
```

Comment:
if it was fully processed, wouldn't it have a result ?

Reply:
I see now that you move the Source back to this state when the result is taken.

I am a little bit on the fence on that one: it clearly works, the question is: are there any situations in which we want to check if a source was already processed?

One alternative way of doing this then, which would make the transitions clear is to have a 4th state:  `Consumed`
takine the result would move it to that state.

It is one more state, but at least the meaning of no state is overloaded.

Reply:
Yeah, this makes sense -- if we are doing the right thing, might do it the right way as well, amended the commit.